### PR TITLE
Workspace and WorkspaceViewModel Serialization

### PR DESCRIPTION
**Pull request by ikeough**
### Purpose
This PR adds two different modes for `Workspace` serialization:
- If serialization is requested from the UI, it goes through `DynamoViewModel` which in turn calls `CurrentSpaceViewModel.Save`. The `WorkspaceViewModel` serialization first calls `WorkspaceModel.ToJson(...)`, then injects the `View` property whose value is the serialization of the `WorkspaceViewModel` into the resulting json. Then the json is saved to the file.
- If serialization is requested using `WorkspaceModel.Save` then `WorkspaceModel.ToJson` is called, and the results are written directly to the file.

A number of API changes have been made and are called out in comments below.

### Declarations

Check these if you believe they are true

- [x] The code base is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [x] User facing strings, if any, are extracted into `*.resx` files
- [x] All tests pass using the self-service CI. (One existing regression remains)
- [ ] Snapshot of UI changes, if any. **n/a**
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions), and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.

### Reviewers

@QilongTang

### CODE REVIEW

File: src/DynamoCore/Graph/Workspaces/SerializationConverters.cs [hunk 1]
```diff
@@ -27,14 +27,14 @@ namespace Dynamo.Graph.Workspaces
     /// These nodes require a CustomNodeDefinition which can only be supplied
     /// by looking it up in the CustomNodeManager.
     /// </summary>
-    public class NodeModelConverter : JsonConverter
+    public class NodeReadConverter : JsonConverter
```
Comment (ikeough) [inline src/DynamoCore/Graph/Workspaces/SerializationConverters.cs]: Potential API Change. Need to check if these converters were included in 1.3.

File: src/DynamoCore/Graph/Workspaces/SerializationExtensions.cs [hunk 2]
```diff
@@ -18,9 +18,7 @@ public static class SerializationExtensions
         /// Save a Workspace to json.
         /// </summary>
         /// <returns>A string representing the serialized WorkspaceModel.</returns>
-        public static string ToJson(this WorkspaceModel workspace, LibraryServices libraryServices,
-            EngineController engineController, DynamoScheduler scheduler, NodeFactory factory,
-            bool isTestMode, bool verboseLogging, CustomNodeManager manager)
+        public static string ToJson(this WorkspaceModel workspace)
```
Comment (ikeough) [inline src/DynamoCore/Graph/Workspaces/SerializationExtensions.cs]: API Change.

File: src/DynamoCore/Graph/Workspaces/WorkspaceModel.cs [hunk 3]
```diff
@@ -143,16 +143,16 @@ internal virtual void OnCurrentOffsetChanged(object sender, PointEventArgs e)
         }
 
         /// <summary>
-        ///     Event that is fired when the workspace is saved.
+        /// Event that is fired when the workspace is saved.
         /// </summary>
-        public event Action WorkspaceSaved;
-        protected virtual void OnWorkspaceSaved()
+        public event Action Saved;
+        internal virtual void OnSaved()
```
Comment (ikeough) [inline src/DynamoCore/Graph/Workspaces/WorkspaceModel.cs]: API changed from `WorkspaceSaved` to `Saved`.

File: src/DynamoCore/Graph/Workspaces/WorkspaceModel.cs [hunk 4]
```diff
@@ -869,22 +869,42 @@ public virtual void Clear()
...
+        /// Workspace's Save method serializes the Workspace to JSON and writes it to the specified file path.
         /// </summary>
-        /// <param name="newPath">The path to save to</param>
-        /// <param name="isBackup">Indicates whether saved workspace is backup or not. If it's not backup,
+        /// <param name="filePath">The path the file.</param>
+        /// <param name="isBackup">A flag indicating whether this save operation represents a backup. If it's not backup,
         /// we should add it to recent files. Otherwise leave it.</param>
-        public virtual bool Save(string newPath, bool isBackup = false)
+        /// <exception cref="ArgumentNullException">Thrown when the file path is null.</exception>
+        public virtual void Save(string filePath, bool isBackup = false)
```
Comment (ikeough) [inline src/DynamoCore/Graph/Workspaces/WorkspaceModel.cs]: API change. `Save` now returns void.

File: src/DynamoCore/Models/DynamoModel.cs [hunk 5]
```diff
@@ -1264,37 +1278,7 @@ public void ForceRun()
         #region save/load
 
         /// <summary>
-        /// Save workspace in Json format to specified path.
-        /// </summary>
-        /// <param name="path">The path to save to</param>
-        /// <param name="ws">workspace to save</param>
-        /// <param name="isBackup">indicate saving for backup</param>
-        public bool SaveWorkspace(string path, WorkspaceModel ws, bool isBackup = false)
```
Comment (ikeough) [inline src/DynamoCore/Models/DynamoModel.cs]: API change. `DynamoModel.SaveWorkspace(...)` has been removed. Use `WorkspaceModel.Save(...)` instead.

File: src/DynamoCoreWpf/ViewModels/Core/SerializationExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    /// SerializationExtensions contains methods for serializing a WorkspaceViewModel to json.
+    /// </summary>
+    public static class SerializationExtensions
+    {
+        /// <summary>
+        /// Serialize the WorkspaceViewModel to JSON.
+        /// </summary>
+        /// <param name="viewModel"></param>
+        /// <returns>A JSON string representing the WorkspaceViewModel</returns>
+        public static string ToJson(this WorkspaceViewModel viewModel)
```
Comment (ikeough) [inline src/DynamoCoreWpf/ViewModels/Core/SerializationExtensions.cs]: API change. `WorkspaceViewModel.ToJson(...)` has been added.

File: src/DynamoCore/Graph/Workspaces/WorkspaceModel.cs [hunk 7]
```diff
@@ -869,22 +869,42 @@ public virtual void Clear()
         }
 
         /// <summary>
-        /// Internal save logic unrelated to serialization.
+        /// Workspace's Save method serializes the Workspace to JSON and writes it to the specified file path.
         /// </summary>
-        /// <param name="newPath">The path to save to</param>
-        /// <param name="isBackup">Indicates whether saved workspace is backup or not. If it's not backup,
+        /// <param name="filePath">The path the file.</param>
```
Comment (QilongTang) [inline src/DynamoCore/Graph/Workspaces/WorkspaceModel.cs]: The path `of` the file

File: src/DynamoCore/Graph/Workspaces/WorkspaceModel.cs [hunk 8]
```diff
@@ -869,22 +869,42 @@ public virtual void Clear()
...
-            if (!isBackup)
+                // Handle Workspace or CustomNodeWorkspace related non-serialization internal logic
+                // Only for actual save, update file path and recent file list
+                if (!isBackup)
+                {
+                    FileName = filePath;
+                    OnSaved();
+                }
+
+                OnSaved();
```
Comment (QilongTang) [inline src/DynamoCore/Graph/Workspaces/WorkspaceModel.cs]: Duplicate `OnSaved()`

File: test/DynamoCoreTests/CoreTests.cs [hunk 9]
```diff
@@ -520,7 +518,7 @@ public void CanSaveAsFileWithNodesInIt()
 
             string fn = "ruthlessTurtles.dyn";
             string path = Path.Combine(TempFolder, fn);
-            CurrentDynamoModel.SaveWorkspace(path, CurrentDynamoModel.CurrentWorkspace);
+            CurrentDynamoModel.CurrentWorkspace.Save(path);
```
Comment (QilongTang) [inline test/DynamoCoreTests/CoreTests.cs]: Neat

File: src/DynamoCore/Graph/Workspaces/WorkspaceModel.cs (hunk 8 above)
Comment (ikeough) [inline src/DynamoCore/Graph/Workspaces/WorkspaceModel.cs]: Good catch. This is a mistake.

File: test/DynamoCoreTests/CoreTests.cs (hunk 9 above)
Comment (ikeough) [inline test/DynamoCoreTests/CoreTests.cs]: This was done to make multiple home workspaces easier in the future.

**Comment by QilongTang** [comment]
Restarting `EngOps Build` to make sure Dynamo builds, other than that the changes looks good

**Comment by QilongTang** [comment]
@ikeough Looks like there are a few minor build errors after merging with master branch:
 Graph\Workspaces\SerializationConverters.cs(203,39): error CS1002: ; expected [E:\Builds\Dynamo_master\Dynamo\src\DynamoCore\DynamoCore.csproj]
  Graph\Workspaces\SerializationConverters.cs(279,38): error CS1002: ; expected [E:\Builds\Dynamo_master\Dynamo\src\DynamoCore\DynamoCore.csproj]

File: src/DynamoCore/Graph/Workspaces/SerializationConverters.cs [hunk 10]
```diff
@@ -177,15 +177,15 @@ public override bool CanWrite
     /// a NodeFactory, and a Scheduler. These must be supplied at the time of 
     /// construction and should not be serialized.
     /// </summary>
-    public class WorkspaceConverter : JsonConverter
+    public class WorkspaceReadConverter : JsonConverter
```
Comment (mjkkirschner) [inline src/DynamoCore/Graph/Workspaces/SerializationConverters.cs]: Hey @ikeough I'm curious what the advantage was for splitting the converters into two?

Comment (ikeough) [inline src/DynamoCore/Graph/Workspaces/SerializationConverters.cs]: @mjkkirschner It's not being split in two. The workspace serialization doesn't need to use this converter. Trying to do so required references to dynamo model objects like the scheduler, which is not allowed in workspace. I removed this converter from the logic in `ToJson(...)`, and changed the name to make it explicit that it's a converter for reading only.

**Pull request merged**